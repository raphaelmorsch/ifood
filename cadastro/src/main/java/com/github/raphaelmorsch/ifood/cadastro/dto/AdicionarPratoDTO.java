package com.github.raphaelmorsch.ifood.cadastro.dto;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import javax.validation.ConstraintValidatorContext;

import com.github.raphaelmorsch.ifood.cadastro.Prato;
import com.github.raphaelmorsch.ifood.cadastro.infra.DTO;
import com.github.raphaelmorsch.ifood.cadastro.infra.ValidDTO;

@ValidDTO
public class AdicionarPratoDTO implements DTO{
	
	public String nome;
	
	public String descricao;
	
	public BigDecimal preco;
	

	@Override
	public boolean isValid(ConstraintValidatorContext constraintValidatorContext) {
		constraintValidatorContext.disableDefaultConstraintViolation();
		Map<String, Object> params = new HashMap<String, Object>();
		params.put("nome", this.nome);
		if (Prato.count("nome = :nome ", params) > 0) {
			constraintValidatorContext.buildConstraintViolationWithTemplate("{prato.existente}").addPropertyNode("name");
			return false;
		}
		return true;
	}

}
