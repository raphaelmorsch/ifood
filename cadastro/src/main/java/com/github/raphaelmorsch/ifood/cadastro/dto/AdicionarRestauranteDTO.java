package com.github.raphaelmorsch.ifood.cadastro.dto;

import javax.validation.ConstraintValidatorContext;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

import org.hibernate.validator.constraints.br.CNPJ;

import com.github.raphaelmorsch.ifood.cadastro.Restaurante;
import com.github.raphaelmorsch.ifood.cadastro.infra.DTO;
import com.github.raphaelmorsch.ifood.cadastro.infra.ValidDTO;

@ValidDTO
public class AdicionarRestauranteDTO implements DTO {

	@NotNull
	public String proprietario;

	@CNPJ
	@NotNull
	public String cnpj;

	@Size(min = 3, max = 30)
	public String nomeFantasia;

	public LocalizacaoDTO localizacao;

	@Override
	public boolean isValid(ConstraintValidatorContext constraintValidatorContext) {
		constraintValidatorContext.disableDefaultConstraintViolation();
		if (Restaurante.find("cnpj", this.cnpj).count() > 0) {
			constraintValidatorContext.buildConstraintViolationWithTemplate("{cnpj.existente}").addPropertyNode("cnpj")
					.addConstraintViolation();
			return false;
		}
		return true;
	}

}
