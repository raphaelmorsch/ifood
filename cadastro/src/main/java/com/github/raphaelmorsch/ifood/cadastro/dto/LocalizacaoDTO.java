package com.github.raphaelmorsch.ifood.cadastro.dto;

public class LocalizacaoDTO {

	public Double longitude;

	public Double latitude;

}
