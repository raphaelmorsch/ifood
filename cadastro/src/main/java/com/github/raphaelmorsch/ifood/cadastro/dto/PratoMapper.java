package com.github.raphaelmorsch.ifood.cadastro.dto;

import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;

import com.github.raphaelmorsch.ifood.cadastro.Prato;

@Mapper(componentModel = "cdi")
public interface PratoMapper {
	
	Prato toPrato(AdicionarPratoDTO dto, @MappingTarget Prato prato);
	
	AdicionarPratoDTO toPratoDTO(Prato entidade);

}
