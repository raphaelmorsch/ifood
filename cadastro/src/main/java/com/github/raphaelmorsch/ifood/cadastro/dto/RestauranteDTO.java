package com.github.raphaelmorsch.ifood.cadastro.dto;

import com.github.raphaelmorsch.ifood.cadastro.Localizacao;

public class RestauranteDTO {
	
	public Long id;

	public String proprietario;

	public String cnpj;

	public String nomeFantasia;

	public String dataCriacao;

	public String dataAtualizacao;

	public Localizacao localizacao;

}
