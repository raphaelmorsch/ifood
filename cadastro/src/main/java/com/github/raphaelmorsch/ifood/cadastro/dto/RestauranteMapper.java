package com.github.raphaelmorsch.ifood.cadastro.dto;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;

import com.github.raphaelmorsch.ifood.cadastro.Restaurante;

@Mapper(componentModel = "cdi")
public interface RestauranteMapper {

	@Mapping(target = "nome", source = "nomeFantasia")
	void toRestaurante(AdicionarRestauranteDTO dto, @MappingTarget Restaurante restaurante);

	@Mapping(target = "nomeFantasia", source = "nome")
	@Mapping(target = "dataCriacao", dateFormat = "dd/MM/yyyy HH:mm:ss")
	RestauranteDTO toRestauranteDTO(Restaurante entity);

}
