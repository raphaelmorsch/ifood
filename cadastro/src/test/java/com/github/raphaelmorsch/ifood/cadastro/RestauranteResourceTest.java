package com.github.raphaelmorsch.ifood.cadastro;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import javax.ws.rs.core.Response.Status;

import org.approvaltests.Approvals;
import org.hamcrest.Matcher;
import org.hamcrest.Matchers;
import org.jose4j.jwt.consumer.InvalidJwtException;
import org.junit.Assert;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.github.raphaelmorsch.ifood.cadastro.dto.AdicionarRestauranteDTO;
import com.github.raphaelmorsch.ifood.cadastro.util.TokenUtils;
import com.google.gson.JsonObject;

import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.http.Header;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;

@DBRider
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE)
@QuarkusTest
@QuarkusTestResource(CadastroTestResourceLifecycleManager.class)
public class RestauranteResourceTest {

	private String token;
	
	@BeforeEach
	public void generateToken() throws InvalidJwtException, Exception {
		this.token = TokenUtils.generateTokenString("/JWTProprietarioClaims.json", null);
	}

	private RequestSpecification given() {

		return RestAssured.given()
				.contentType(ContentType.JSON).accept(ContentType.ANY)
				.header(new Header("Authorization", "Bearer " + token));
	}

	@Test
	@DataSet("restaurantes-cenario-1.yml")
	public void testBuscarRestaurantes() {
		String resultado = given().when().get("/restaurantes").then().statusCode(200).extract().asString();
		Approvals.verifyJson(resultado);
	}

	@Test
	public void testAdicionarRestaurante() {
		Map<String, Matcher<?>> responseHeaders = new HashMap<String, Matcher<?>>();
		responseHeaders.put("content-length", Matchers.equalTo("0"));
		responseHeaders.put("location", Matchers.equalTo("http://localhost:8081/restaurantes/1"));

		Map<String, String> requestHeaders = new HashMap<String, String>();
		requestHeaders.put("accept", "*/*");
		requestHeaders.put("Content-Type", "application/json");

		JsonObject jsonRequest = new JsonObject();
		jsonRequest.addProperty("cnpj", "58.888.082/0001-32");
		jsonRequest.addProperty("nome", "Meu Novo Restaurante");
		jsonRequest.addProperty("proprietario", "Novo ID do keycloak");

		given().when().headers(requestHeaders).body(jsonRequest.toString())
				.post("/restaurantes", Collections.emptyMap()).then().assertThat().headers(responseHeaders)
				.statusCode(Status.CREATED.getStatusCode());
	}

	@Test
	@DataSet("restaurantes-cenario-1.yml")
	public void atualizarRestaurante() {

		AdicionarRestauranteDTO dto = new AdicionarRestauranteDTO();
		dto.nomeFantasia = "novo nome";
		dto.cnpj = "57.218.236/0001-16";
		dto.proprietario = "novo proprietario";
		Long parameterValue = 123L;
		
		Response resposta = given().when().body(dto).put("/restaurantes/{id}", parameterValue);

		resposta.then()
				.statusCode(Status.NO_CONTENT.getStatusCode()).and()
				.header("content-location", Matchers.endsWith("/restaurantes/" + parameterValue.toString()));

		Restaurante findById = Restaurante.findById(parameterValue);
		Assert.assertEquals(dto.nomeFantasia, findById.nome);

	}

}