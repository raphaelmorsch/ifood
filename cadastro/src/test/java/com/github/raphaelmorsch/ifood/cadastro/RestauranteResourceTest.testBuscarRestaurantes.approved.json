[
  {
    "cnpj": "97.791.505/0001-53",
    "id": 123,
    "nomeFantasia": "Novo Restaurante 1",
    "proprietario": "Id do Keycloak 1"
  },
  {
    "cnpj": "15.188.530/0001-82",
    "id": 234,
    "nomeFantasia": "Novo Restaurante 2",
    "proprietario": "Id do Keycloak 2"
  }
]